8288838: jpackage: file association additional arguments

akashche opened the pull request:
jpackage implementation of file association on Windows currently passes a selected filename as an only argument to associated executable.

It is proposed to introduce additional option in file association property file to allow optionally support additional arguments using `%*` batch wildcard.

Note, current implementation, while fully functional, is only a **DRAFT** one, it is not ready for integration in this form. I would appreciate any guidance on the following points:

 - option naming inside a properties file, currently `pass-all-args` is used
 - option naming in a bundler parameter implementation, it is not clear if it should introduce a new group of "file association windows specific options" next to the existing "file association mac specific options" group
 - test organization to cover the new option: currently it is included inside `FileAssociationTest` and piggybacks on the existing (and unrelated) `includeDescription` parameter; it is not clear whether it should be done in a separate test and whether to include runs for every parameter combination
 - test run implementation: currently arguments are checked when a file with associated extension is invoked from command line; it is not clear whether it would be more appropriate instead to create a desktop shortcut with the same command as a target and to invoke it with `java.aws.Desktop`

Also please note, that full install/uninstall run is currently enabled in `FileAssociationTest`, it is intended to be used only in a draft code during the development and to be removed (to use the same "install or unpack" logic as other tests) in a final version.

Testing:
 
- [x] test to cover new logic is included
- [x] ran jtreg:jdk/tools/jpackage with no new failures

akashche commented on src/jdk.jpackage/windows/classes/jdk/jpackage/internal/WixAppImageFragmentBuilder.java:
```diff
@@ -529,7 +529,11 @@ private List<String> addFaComponents(XMLStreamWriter xml,
                 xml.writeAttribute("Command", "Open");
-                xml.writeAttribute("Argument", "\"%1\"");
+                if (fa.passAllArguments) {
+                    xml.writeAttribute("Argument", "\"%1\" %*");
```
I've added argument with spaces to the test, they seem to be passed correctly (with existing unquoted `%*`) with both command line invocation and with a shortcut.

akashche commented:
> What would be the alternative to `pass-all-args`? 

As this is intended to be a public property, I assume there should be a CSR and some consistent name, perhaps with `win-` prefix - I assume we will need an additional similar property for a verb label displayed in a context menu. `pass-all-args` was intended only as a working title.

> What if we unconditionally add `%*` batch wildcard?

Can we change the existing default behaviour? Should we? It won't break existing usage, but is it actually always desired to support passing arbitrary arguments from the desktop shortcut to the target java app? I myself cannot answer any of these questions.

alexeysemenyukoracle commented:
> I assume we will need an additional similar property for a verb label displayed in a context menu.

Agree. At least we need to provide l10n for it.

> Can we change the existing default behaviour? Should we? It won't break existing usage, but is it actually always desired to support passing arbitrary arguments from the desktop shortcut to the target java app? I myself cannot answer any of these questions.

The existing behavior simply cuts off all, but the first argument from the argument array passed to the app launcher. Changing this looks more like a fix of a bug, than an enhancement. 
I'd keep it simple: replace the value of the `Argument` attribute from `%1` to `%*`. Besides, `"%1 %*"` seems to duplicate the first argument (`%1` is not referencing a path to app launcher executable, but the name of a file to open).
If fixed in this way, we don't need to introduce `pass-all-args`.

akashche commented:
> The existing behavior simply cuts off all, but the first argument from the argument array passed to the app launcher. Changing this looks more like a fix of a bug, than an enhancement. I'd keep it simple: replace the value of the `Argument` attribute from `%1` to `%*`. Besides, `"%1 %*"` seems to duplicate the first argument (`%1` is not referencing a path to app launcher executable, but the name of a file to open). If fixed in this way, we don't need to introduce `pass-all-args`.

I've experimented with this, used following example app:

 - `hello.jar` - java application
 - `hello.exe` - launcher created by jpackage
 - `hello.foo` - file with a `.foo` extension associated with `hello.exe`

Arguments placeholders specified In a `progid[...]\shell\open\command` registry entry are received in java app in a following form:

 - `%0` points to `c:\path\too\hello.foo`, that is a selected file path (for context menu invocation) or a `shortcut.TargetPath` property (for `LNK` shortcut invocation)
 - `%1` is the same as `%0` 
 - `%2` points to the first argument in `shortcut.Arguments` property
 - `%*` contains all the arguments beginning from `%2`

`%1` must be quoted, otherwise it only contains the part of path before the first space and other path parts are passed as `%2` and following arguments.

`%*` must not be quoted, it passes arguments with spaces correctly.

Based on the above I've updated the patch dropping the property and changing the `Argument` to `"%1" %*`.

I would appreciate the guidance on the test changes (note, `packageTest.forTypes` and `packageTest.run` changes are temporary). Foremost, should we cover all three ways to invoke the associated file (command line, desktop open associated file, desktop open LNK) and (if yes) should they be split into separate test runs?

alexeysemenyukoracle commented:
> For runtime installers it would be great to have a custom verb label, to be able to use something like: "Open with VendorName OpenJDK". I intend to prototype this and file in a separate PR

Sounds good.

> %* contains all the arguments beginning from %2

Oh, interesting. I didn't know that! Thank you for providing a detailed explanation.

> I would appreciate the guidance on the test changes

I put proposed changes to test classes in https://github.com/openjdk/jdk/pull/9331 PR. I hope the approach gives enough flexibility to configure testing of fa. We can configure all possible test configurations in `FileAssociationTest` test and have basic testing of fa in tests where it is not primary focus. Code in https://github.com/openjdk/jdk/pull/9331 is a sketch, it compiles, but I didn't run it.

akashche commented:
I've rebased the changes on top of PR 9331 and rebased the result on top of recent master. I think the patch if ready for review now.

Note on non-ASCII arguments support: I've found that Windows system locale needs to be changed to support specific language. Checked manually that such args are passed successfully with both command-line and LNK shortcut arguments. Left a note on this in test.

alexeysemenyukoracle commented:
Looks good! I'll try the patch locally and get back to you.
Please provide a detailed test scenario for non-ASCII args. We can add this scenario to manual test runs done by SQE.

akashche commented:
About the non-ASCII args: in `FileAssociations.TestRun::openFiles` non-ASCII argument can be supplied along with other arguments (in `WinCommandLine` and `WinDesktopOpenShortcut` runs). The arguments passing and the check in the output file works correctly for such arguments, but only when the `Language for non-Unicode programs` system setting in Windows is set to the corresponding language. Otherwise the argument is received by java app in `?????????` form and the test expectedly fails. I've checked that this behaviour of jpackage launcher is the same as in the main `java.exe` launcher. I've added the note comment and the concrete example of an arg to `FileAssociations.TestRun::openFiles`.

If it is desired to be able to run such non-ASCII test manually (without changing the test code), I assume this can be done by making the test to read arguments from filesystem, or to add some kind of a test option to run the branch with predefined non-ASCII arguments only when requested.

alexeysemenyukoracle commented:
I confirm the patch works and there are no regressions.

SQE uses only the part of jpackage jtreg tests that create test packages, so there is no point in customizing tests with reading args from the file system. I think I have enough information to help SQE update test spec and cover Unicode args.

alexeysemenyukoracle commented:
/sponsor

akashche commented on test/jdk/tools/jpackage/share/FileAssociationsTest.java:
```diff
@@ -106,7 +105,8 @@ public static void test(boolean includeDescription) {
                 .applyTo(packageTest);
 
-        packageTest.run();
+        packageTest.run(RunnablePackageTest.Action.CREATE, RunnablePackageTest.Action.INSTALL,
```
Thanks for the info! I've seen `jpackage.test.action` property in `run_tests.sh`, but never used it.

akashche commented:
/integrate

akashche commented:
Thanks for the reviews!

alexeysemenyukoracle commented:
/sponsor
